src/zcash/IncrementalMerkleTree.hpp

```diff
@@ -57,25 +58,36 @@ class MerklePath {
 template<size_t Depth, typename Hash>
 class EmptyMerkleRoots {
 public:
-    EmptyMerkleRoots() {
-        empty_roots.at(0) = Hash::uncommitted();
-        for (size_t d = 1; d <= Depth; d++) {
-            empty_roots.at(d) = Hash::combine(empty_roots.at(d-1), empty_roots.at(d-1), d-1);
-        }
+    void initialize() {
+        if (!initialized) {
+     LOCK(cs_EmptyMerkleRoots_private_variables_write);
+          if (!initialized) {
+            empty_roots.at(0) = Hash::uncommitted();
+            for (size_t d = 1; d <= Depth; d++) {
+                empty_roots.at(d) = Hash::combine(empty_roots.at(d-1), empty_roots.at(d-1), d-1);
+            }
+            initialized = true;
+          }
+       }
     }
+    EmptyMerkleRoots() { initialized = false; }
     Hash empty_root(size_t depth) {
+        initialize();
         return empty_roots.at(depth);
     }
     template <size_t D, typename H>
     friend bool operator==(const EmptyMerkleRoots<D, H>& a,
                            const EmptyMerkleRoots<D, H>& b);
 private:
+    CCriticalSection cs_EmptyMerkleRoots_private_variables_write;
+    bool initialized;
     std::array<Hash, Depth+1> empty_roots;
 };
 
 template<size_t Depth, typename Hash>
 bool operator==(const EmptyMerkleRoots<Depth, Hash>& a,
                 const EmptyMerkleRoots<Depth, Hash>& b) {
+    a.initialize(); b.initialize();
```

Comment:
This did not compile for me, because a and b are declared `const`, but `initialize()` modify them. Does this compile for you? In my local repo, I removed both of the `const` qualifiers (from here and above at lines 79 and 80), and it compiles. I think that's all you can really do to fix this, because if you try to fake it (since from the caller's perspective, `==` shouldn't require the arguments to be mutable), I think that's confusing, and potentially unsafe.

Reply:
Yep, as it stands the compiler would be free to move the comparison *before* the calls to `initialize`. That's what `const` is for, to allow "optimisations" like that.

Reply:
A similar but more subtle problem occurs in `IncrementalMerkleTree::operator==`. Because `IncrementalMerkleTree` includes an instance of `EmptyMerkleRoots` *by value*, and the former is const in that method, the compiler is allowed to assume that the instance of the latter is const, even if not explicitly recalled as such in `EmptyMerkleRoots::operator==`.

Reply:
(`const` should probably get some kind of award for most misunderstood language "feature". It's a "please shoot me in the foot now" feature, not a safety feature.)

Reply:
Thanks for the review, but unfortunately I haven't understood from it what you want me to do.

Reply:
Another option might be to declare `initialize()` as `const` and the fields of `EmptyMerkleRoots` as mutable.

Reply:
So now the compiler won't move the comparison above the calls to `initialize()`, even if we make `initialize()` const, because it knows that `EmptyMerkleRoots` has mutable members?